src/Sarif/Writers/SarifConsolidator.cs

```diff
@@ -0,0 +1,259 @@
+                    newLocations.Add(loc);
+                }
+            }
+
+            return newLocations;
+        }
+
+        public void Trim(Location loc)
+        {
+            if (loc != null)
+            {
+                // Remove Ids
+                loc.Id = -1;
+
+                Trim(loc.PhysicalLocation);
+                Trim(loc.LogicalLocation);
+                Trim(loc.LogicalLocations);
+            }
+        }
+
+        public IList<LogicalLocation> Trim(IList<LogicalLocation> locations)
+        {
+            if (locations == null || locations.Count == 0) { return locations; }
+
+            HashSet<LogicalLocation> uniqueLocations = new HashSet<LogicalLocation>(LogicalLocation.ValueComparer);
+            List<LogicalLocation> newLocations = new List<LogicalLocation>();
+
+            this.TotalLocations += locations.Count;
+            for (int i = 0; i < locations.Count; ++i)
+            {
+                LogicalLocation loc = locations[i];
+
+                // Trim Locations themselves
+                Trim(loc);
+
+                // Keep only one copy of each unique location
+                if (uniqueLocations.Add(loc))
+                {
+                    this.UniqueLocations++;
+                    newLocations.Add(loc);
+                }
+            }
+
+            return newLocations;
+        }
+
+        public void Trim(LogicalLocation loc)
+        {
+            if (loc != null)
+            {
+                // Leave only index if index is present
+                if (loc.Index != -1)
+                {
+                    loc.DecoratedName = null;
+                    loc.FullyQualifiedName = null;
+                    loc.Name = null;
+                    loc.ParentIndex = -1;
+                    loc.Properties?.Clear();
+                    loc.Tags?.Clear();
+                }
+            }
+        }
+
+        public void Trim(PhysicalLocation loc)
+        {
+            if (loc != null)
+            {
+                Trim(loc.ArtifactLocation);
+                Trim(loc.Region);
+                Trim(loc.ContextRegion);
+            }
+        }
+
+        public void Trim(ArtifactLocation a)
+        {
+            if (a != null)
+            {
+                // Leave only index if index is present
+                if (a.Index != -1)
+                {
+                    a.Uri = null;
+                    a.UriBaseId = null;
+                    a.Description = null;
+                    a.Tags?.Clear();
+                    a.Properties?.Clear();
+                }
+            }
+        }
+
+        public void Trim(Region r)
+        {
+            if (r != null && r.StartLine > 0)
+            {
+                // Remove EndLine if the same as StartLine
+                if (r.EndLine == r.StartLine)
+                {
+                    r.EndLine = 0;
+                }
+
+                // Remove Offset/Length if Line/Column available
```

Comment:
I actually think that in the spirit of favoring compactness, it would have been more consistent for us to zap the line/column details when offset/length are available. The 'readable' mode would zap offset/length instead.

---
In reply to: [323970078](https://github.com/microsoft/sarif-sdk/pull/1672#discussion_r323970078) [](ancestors = 323970078)

Reply:
By the way, the start/end line/column stuff has very elaborate conventions for defaulting various values. If start and end line are identical, for example, you don't need to recapitulate the end line. If the endColumn happens to be the last char in the line, you don't need that (which is not a condition that can be computed strictly from the SARIF log file, however). You could consider reviewing these rules and updating Trim(Region) based on them (at least capturing some of the more basic ones).

---
In reply to: [324282310](https://github.com/microsoft/sarif-sdk/pull/1672#discussion_r324282310) [](ancestors = 324282310,323970078)

Reply:
I added one more rule - remove StartColumn when it's 1. The other rules can't be evaluated without the file text, which would be really expensive to evaluate in this context.